Tide makes decision based on common code review struct instead of directly accessing PullRequest

chaodaiG opened the pull request:
This is a refactoring, in preparation for tide supporting other code review providers, such as gerrit.

There were two possible approaches:
- a). Interfaces. For instance defining an interface supporting shared functions such as `GetNumber()`, `GetOrg()`, `GetRepo()`. And any GitHub only fields will be under an another interface, for example:
```
type CodeReviewCommon interface {
  GetNumber()

  GitHubOnly
}

type GitHubOnly interface {
  GetMilestone() *Milestone
}
```

- b). Common struct. The fields of this struct are derived directly from GitHub struct and potentially future struct such as from Gerrit.

I have almost finished the implementation of a), and while fixing the last failed unit test, realized that the `PullRequest` struct is marshalled by tide and unmarshalled by deck as part of `pool`, the use of interface made the case harder. Potentially we could implement our own unmarshaller, but the problem is in typescript, which would not agree with the unmarshalling logic. So implementing with b) feels like a cleaner solution.

/cc @cjwagner @alvaroaleman @mpherman2 @listx 
for early feedbacks. Note that this PR is missing the typescript change, which I'll work on early next week

chaodaiG commented:
Updated typescript as well, this PR is now ready for review.
@cjwagner, @mpherman2 , @listx, @alvaroaleman please take a look

mpherman2 commented on prow/tide/status.go:
```diff
@@ -423,7 +423,7 @@ func (sc *statusController) setStatuses(all []PullRequest, pool map[string]PullR
 		}
 		actualState = githubql.StatusState(strings.ToLower(string(actualState)))
-		if !sc.dontUpdateStatus.has(string(pr.Repository.Owner.Login), string(pr.Repository.Name), int(pr.Number)) && (wantState != string(actualState) || wantDesc != actualDesc) {
+		if !sc.dontUpdateStatus.has(pr.Org, pr.Repo, pr.Number) && (wantState != string(actualState) || wantDesc != actualDesc) {
```
Does `string(pr.Repository.Owner.Login` translate to `pr.Org` here? Seems like this should be pr.AuthorLogin. Noticed this a few places below too.

chaodaiG replied:
Yeah it actually does. Imo for GitHub:
- `.Repository` is the target repo, so `.Repository.Owner.Login` means the target org, in this PR `.Repository` is `kubernetes/test-infra`, and `.Repository.Owner.Login` is `kubernetes`.
- `.AuthorLogin` is the source repo, in this PR it's `chaodaiG`

mpherman2 commented on prow/tide/tide.go:
```diff
@@ -726,7 +726,7 @@ func (m *mergeChecker) isAllowed(pr *PullRequest) (string, error) {
 		return "PR can't be rebased", nil
 	}
-	orgRepo := config.OrgRepo{Org: string(pr.Repository.Owner.Login), Repo: string(pr.Repository.Name)}
+	orgRepo := config.OrgRepo{Org: pr.Org, Repo: pr.Repo}
```
Same here?

chaodaiG replied:
see my other comment

mpherman2 commented on prow/tide/codereview.go:
```diff
@@ -0,0 +1,96 @@
+/*
```
Are we planning on only using this for Tide? If not is this still the place we want to keep this?

chaodaiG replied:
At least for this work(gerrit tide) this is only used for Tide. I think it makes more sense to keep it here in this case, to avoid confusion

alvaroaleman commented on prow/tide/codereview.go:
```diff
@@ -0,0 +1,96 @@
+	Org    string
+	Repo   string
+	// BaseRefPrefix gets prefix of ref, such as /refs/head, /refs/tags
+	BaseRefPrefix string
```
This seems to be unused

chaodaiG replied:
https://github.com/kubernetes/test-infra/blob/b0c002975451374244eb4b8a5ec1aec965bc98f4/prow/tide/tide.go#L1869 it's actually used

alvaroaleman commented on prow/tide/tide.go:
```diff
@@ -835,36 +835,36 @@ func unsuccessfulContexts(contexts []Context, cc contextChecker, log *logrus.Ent
 	prLabels := sets.NewString()
-	for _, l := range pr.Labels.Nodes {
-		prLabels.Insert(string(l.Name))
+	for _, l := range pr.Labels {
+		prLabels.Insert(l)
```
Nit:
```suggestion
        prLabels := sets.NewString(pr.Labels)
```

alvaroaleman commented on prow/tide/codereview.go:
```diff
@@ -0,0 +1,96 @@
+	}
+}
+
+func CodeReviewCommonFromPullRequest(pr *PullRequest) *CodeReviewCommon {
```
I think it would be good to test this function and to do it in a way where the test will fail if add new fileds and forget to update this. The best I can think of to make that happen would be to also build the reverse (`CodeReviewToPullRequest`) and then use fuzzed data and check if we get the same `PullRequest` after a roundtrip to `CodeReviewCommon` and back

chaodaiG replied:
Totally agreed. I have done an experiment earlier with `PullRequest` and `gerrit.ChangeInfo` as substructs under `CommonCodeReview` struct, fuzzing `PullRequest` works as expected, but `gerrit.ChangeInfo` takes forever to fuzz and timed out after 30s, so we will run into trouble with gerrit with this roundtrip test. Will need to think a little more on how to make this work

alvaroaleman commented on prow/tide/codereview.go:
```diff
@@ -0,0 +1,96 @@
+	AuthorLogin string
+	// Labels gets labels on the PR.
+	// TODO(chaodaiG): labels might mean something different on gerrit, consider
+	// how to name this nicely.
```
IME when trying to different underlying apis away they will always end up having subtle but important differences, even if they look very similiar. So while it seems ok to me to to have a common `data` type, the code that actually interprets this should be seperated or it will end up being riddled with `if isGithub {} else {}` sooner or later. This doesn't only apply to the obviously provider-dependent parts like say `search` but also to something like `filterSubpool` (i.E. there should be one for GitHub, one for Gerrit. They might end up sharing code but should be kept distinct).

This in turn also means that we can be more explicit here and don't need to to try equalize GitHub and Gerrit. I don't know Gerrit very well, but in doubt we could add something like `GerritLabels`.

Writing this out, I am starting to wonder if it would be useful to use substructs, i.E. comprise this out of a `CommonCodeReview`, `GitHubCodeReview`, `GerritCodeReview` struct to more clearly denoate what concepts apply where. If we embedd the substructs and use `jsoin:",inline`, this won't affect serialization and thus won't break Deck.

alvaroaleman replied:
Or maybe the `common` part doesn't need to be a substruct, but moving the fields that are provider-specific into their own substruct would be nice and make it more clear what fields can be assumed to be set when.

chaodaiG replied:
This is a good point. Had an offline discussion with Cole and we both liked this idea, with concern of potential collisions between embedded structs. So ran an experiment https://go.dev/play/p/vp502rgfVpN, TLDR is that if there is any shared field between two embedded structs, yaml unmarshal just discards the field instead of assigning the value to both, so there is potential risk of certain fields getting lost during repeated marshal/unmarshal. The safest way to do so is probably substruct like:
```
type CommonCodeReview struct {
  // Common fields consumed by Deck
  Org string
  // Provider specific
  GitHub *GitHubCodeReview
  Gerrit *GerritCodeReview
}
```

Also agree with the concern over `if isGithub {} else if isGerrit {}`, I think we can try to cut a line that if we ever need to do this then the field should not be a common field. For instance `Labels` is not a common field.

chaodaiG commented on prow/cmd/deck/static/api/pr.ts:
```diff
@@ -5,9 +5,29 @@ export interface Label {
 export interface PullRequest extends BasePullRequest {
   Merged: boolean;
   Title: string;
+  Author: Author;
```
Note that there are two different `PullRequest` definition in prow, one for tide https://github.com/kubernetes/test-infra/blob/67b179342c2a6bd7a447b0e70cb94cf6d7fe0f55/prow/tide/tide.go#L1909 and one for the rest:https://github.com/kubernetes/test-infra/blob/67b179342c2a6bd7a447b0e70cb94cf6d7fe0f55/prow/prstatus/prstatus.go#L87

The same happens in Typescript with inheritance as you have noted here. The fields that are moved out of BasePullRequest is that they are not really used by tide at all, and doesn't make sense to be kept in Base struct, typescript compilation would catch error if this field is later on used. 

This might be a tech debt, but might not be an easy one to solve. I'd prefer not to fix this "tech debt" right now due to the time pressure of what I need to deliver

chaodaiG commented on prow/tide/codereview.go:
```diff
@@ -0,0 +1,96 @@
+
+func CodeReviewCommonFromPullRequest(pr *PullRequest) *CodeReviewCommon {
+	crc := &CodeReviewCommon{
+		NameWithOwner:  string(pr.Repository.NameWithOwner),
```
good catch, `RepoWithOwner` is now gone

chaodaiG commented:
Updated the PR with substruct implementation. Other things worth noting:
- Introduced `CodeReviewForDeck` struct, which is the required info by deck from tide.
- Introduced `type MinCodeReviewCommon CodeReviewCommon` with custom `Marshal` function, which allows tide to only send relevant fields to deck instead of full PullRequest struct. (As mentioned earlier, that the full struct of gerrit ChangeInfo is very likely way larger than github PullRequest struct, so this is almost a must to do). On the other hand, this also makes it more clearly what exactly deck wants.

@cjwagner @alvaroaleman please take another look
/hold
Friday might not be the best day to merge this PR

cjwagner commented on prow/tide/tide.go:
```diff
@@ -575,9 +629,9 @@ func (c *Controller) initSubpoolData(sp *subpool) error {
 	}
 	sp.cc = make(map[int]contextChecker, len(sp.prs))
 	for _, pr := range sp.prs {
-		sp.cc[int(pr.Number)], err = c.config().GetTideContextPolicy(c.gc, sp.org, sp.repo, sp.branch, refGetterFactory(string(sp.sha)), string(pr.HeadRefOID))
+		sp.cc[pr.Number], err = c.config().GetTideContextPolicy(c.gc, sp.org, sp.repo, sp.branch, refGetterFactory(string(sp.sha)), pr.HeadRefOID)
 		if err != nil {
-			return fmt.Errorf("error setting up context checker for pr %d: %w", int(pr.Number), err)
+			return fmt.Errorf("error setting up context checker for pr %d: %w", pr.Number, err)
```
FWIW we'll probably need to include the initSubpoolData function in the controller interface since this will be different for Gerrit and GitHub. The use of the `sp.presubmits` map could be consistent between the two, but the generation of the map (via `presubmitsByPull`) will likely need to consider Gerrit reporting labels and the context policy checker only applies to GitHub.

cjwagner commented on prow/tide/tide.go:
```diff
@@ -610,7 +664,7 @@ func filterSubpool(ghc githubClient, mergeAllowed func(*PullRequest) (string, er
 //   'pending' because this prevents kicking PRs from the pool when Tide is
 //   retesting them.)
-func filterPR(ghc githubClient, mergeAllowed func(*PullRequest) (string, error), sp *subpool, pr *PullRequest) bool {
+func filterPR(ghc githubClient, mergeAllowed func(*CodeReviewCommon) (string, error), sp *subpool, pr *CodeReviewCommon) bool {
```
As with initSubpoolData, I expect the filterPR function to be a function in the controller interface due to the differences between Gerrit and GitHub.

cjwagner commented on prow/tide/tide.go:
```diff
@@ -714,19 +771,26 @@ func (m *mergeChecker) repoMethods(orgRepo config.OrgRepo) (map[types.PullReques
 // allowed merge method. If there is no error it returns a string explanation if
 // not allowed or "" if allowed.
-func (m *mergeChecker) isAllowed(pr *PullRequest) (string, error) {
+func (m *mergeChecker) isAllowed(crc *CodeReviewCommon) (string, error) {
```
If the `mergeChecker` is GitHub specific then this should just take PullRequest. Or is the plan to make this an interface?

cjwagner replied:
Either way, `prMergeMethod` only makes sense for GitHub PRs so that should just take a PullRequest.

cjwagner commented on prow/tide/search.go:
```diff
@@ -40,7 +40,7 @@ func floor(t time.Time) time.Time {
 }
 
-func search(query querier, log *logrus.Entry, q string, start, end time.Time, org string) ([]PullRequest, error) {
+func search(query querier, log *logrus.Entry, q string, start, end time.Time, org string) ([]CodeReviewCommon, error) {
```
Since this is a GitHub specific search function I would have expected this to return a slice of PullRequests and for that slice to be converted separately, but if this is a more convenient place to do the change that seems fine.

cjwagner reviewed (commented):
I have some more thoughts under the umbrella of separating the shared and GitHub specific logic, but I think avoiding the use of the CodeReviewCommon struct in GitHub specific contexts will resolve most of that and make the boundary between the two areas clearer.

chaodaiG commented on prow/tide/tide.go:
```diff
@@ -1122,9 +1189,10 @@ func (c *Controller) pickBatch(sp subpool, cc map[int]contextChecker, newBatchFu
 	for _, pr := range sp.prs {
-		if c.isRetestEligible(sp.log, &pr, cc[int(pr.Number)]) {
+		// isRetestEligible alters the input, so make a copy
+		if c.isRetestEligible(sp.log, pr.Clone(), cc[pr.Number]) {
```
yes agreed. This is due to a "bug" that should cache prefetched HEAD OID but didn't. Since this PR is already huge I'm not sure that I should introduced the fix to this bug here. So addressing in https://github.com/kubernetes/test-infra/pull/26375

chaodaiG replied:
Now that #26375 is merged, I'll rebase this PR and remove this `Clone` function

chaodaiG commented:
Rebased on top of #26375

chaodaiG commented on prow/cmd/deck/main.go:
```diff
@@ -1284,10 +1286,14 @@ func handleTidePools(cfg config.Getter, ta *tideAgent, log *logrus.Entry) http.H
 
+		var poolsForDeck []tide.PoolForDeck
+		for _, pool := range pools {
+			poolsForDeck = append(poolsForDeck, *tide.PoolToPoolForDeck(&pool))
```
Note that `/tide.js` endpoint is served by deck instead of tide. So the struct is actually minified before consumed by the other pages on prow UI

cjwagner replied:
Yes that makes sense to me, but I'm wondering why we are only minifying the data served by Deck and still sending the unminified data from Tide to Deck. I think we could do minify earlier here instead: https://github.com/kubernetes/test-infra/blob/845379f573ce2b34c3f5c2f2415aea417d22413d/prow/tide/tide.go#L503-L514

chaodaiG commented:
Agree with @cjwagner 's comments about using `PullRequest` struct in logics that are only applicable to GitHub instead of referencing `CodeReviewCommon` everywhere, and this is exactly what I'm planning to do on the next PR.

My original plan is to break the refactoring of tide into two PRs:
- First PR replacing the usage of `PullRequest` struct with the generic `CodeReviewCommon` struct. Will try my best not to change the original logic, so that the PR is easier to be reviewed, and less risk of breaking tide.
- Second PR further separates the logic of GitHub related logics, in which the GitHub related logics will be abstracted into interface(s).

I have worked on the second PR for a bit and realized that it's not a very straightforward change. So might prefer to merge current PR as is, and I can create another branch to keep working on top of this PR.

@cjwagner WDYT?

cjwagner commented on prow/tide/codereview.go:
```diff
@@ -0,0 +1,173 @@
+	IsMergeable   bool
+	UpdatedAtTime time.Time
+
+	Mergeable    string
```
It looks like `IsMergeable bool` should be removed, it is not used anywhere and is redundant.

cjwagner reviewed (approved):
Merging this PR and iterating to address separating the GitHub specific logic sounds fine to me, but could you fix the smaller nits or bugs first please?
Namely the `IsMergeable` field appears redundant and unused.

chaodaiG commented:
> Merging this PR and iterating to address separating the GitHub specific logic sounds fine to me, but could you fix the smaller nits or bugs first please? Namely the `IsMergeable` field appears redundant and unused.

removed.

chaodaiG commented:
/unhold
